executor/aggregate.go

```diff
@@ -38,24 +39,96 @@ type HashAggExec struct {
 	aggCtxsMap    aggCtxsMapper
 	groupMap      *mvmap.MVMap
 	groupIterator *mvmap.Iterator
+	inputRow      chunk.Row
 	GroupByItems  []expression.Expression
 }
 
 // Close implements the Executor Close interface.
 func (e *HashAggExec) Close() error {
+	if err := e.baseExecutor.Close(); err != nil {
+		return errors.Trace(err)
+	}
 	e.groupMap = nil
 	e.groupIterator = nil
 	e.aggCtxsMap = nil
-	return errors.Trace(e.children[0].Close())
+	return nil
 }
 
 // Open implements the Executor Open interface.
 func (e *HashAggExec) Open(goCtx goctx.Context) error {
+	if err := e.baseExecutor.Open(goCtx); err != nil {
+		return errors.Trace(err)
+	}
 	e.executed = false
 	e.groupMap = mvmap.NewMVMap()
 	e.groupIterator = e.groupMap.NewIterator()
 	e.aggCtxsMap = make(aggCtxsMapper, 0)
-	return errors.Trace(e.children[0].Open(goCtx))
+	e.inputRow = e.childrenResults[0].End()
+	return nil
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *HashAggExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	// In this stage we consider all data from src as a single group.
+	if !e.executed {
+		err := e.innerNextChunk(goCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if (e.groupMap.Len() == 0) && len(e.GroupByItems) == 0 {
+			// If no groupby and no data, we should add an empty group.
+			// For example:
+			// "select count(c) from t;" should return one row [0]
+			// "select count(c) from t group by c1;" should return empty result set.
+			e.groupMap.Put([]byte{}, []byte{})
+		}
+		e.executed = true
+	}
+	chk.Reset()
+	for {
+		if chk.NumRows() == e.ctx.GetSessionVars().MaxChunkSize {
+			return nil
+		}
+		groupKey, _ := e.groupIterator.Next()
+		if groupKey == nil {
+			return nil
+		}
+		aggCtxs := e.getContexts(groupKey)
+		for i, af := range e.AggFuncs {
+			af.SetResultInChunk(chk, i, aggCtxs[i])
+		}
```

Comment:
Consider a query:
```sql
create table t(a bigint, b bigint);
insert into t values(1, 1), (2, 2);
select 10 from t group by a;
```
In this case, `len(e.AggFuncs)` is equal to `0`, witch means `e.Schema().Len()` is equal to `0` as well, we should consider `chk.SetNumVirtualRows()` in this case